### DISCUSSION TITLE
Fetching items from an API

### DISCUSSION THREAD

**Discussion by faressoft** [LaunchMenu/LaunchMenu#186] [Q&A] [answered]
I want to achieve this auto-completion:

- Take the search query from LaunchMenu
- Pass it to a RESTful API that filters by "?query=test"
- Get the already filtered and sorted results and display them as is
- Update the list every time the user changes the query by calling the endpoint
- Allow the user to select one of these items and then show another list of actions, such as opening a URL.

Any hints on how I can do this with LaunchMenu?

```js
export default declare({
  info,
  settings,
  open({ context, onClose }) {
    // The code should be written here after the aplet is opened
  },
});
```

**Comment by TarVK** [comment]
This functionality is quite similar to that in the dictionary. 
We wanted to have some helper functions for this eventually, since it's a pretty common case, but there aren't any yet.

For now you can take inspiration from the dictionary applet's code: [packages/applets/dictionary](https://github.com/LaunchMenu/LaunchMenu/blob/development/packages/applets/dictionary/src/index.tsx)

You can also look at the very simple example here: [examples/demosSearch](https://github.com/LaunchMenu/LaunchMenu/blob/development/examples/demosSearch/src/index.tsx)

Unfortunately, making the list show up as ordered by the API isn't as easy. All the default behaviors make sure to apply a search prioritization on it later, to allow different applets to show their results appropriately. In order to overwrite this behavior, you can do something similar to seen here: [makeItemAlwaysShow.ts](https://github.com/LaunchMenu/LaunchMenu/blob/development/packages/core/src/actions/types/search/tracedRecursiveSearch/utils/makeItemAlwaysShow.ts#L34)
Basically you take out the search-action-binding from the menu items you create, using the `adjustBindings` function with a filter, and then create a new binding for `tracedRecursiveSearchHandler` (the search-action handler which also adds a trace to the path it originates from) where you manually specify the [priority](https://launchmenu.github.io/docs/concepts/interaction/search-system#priorities). 

Finally, a list of actions can be added to the context menu. This also works through the [action system](https://launchmenu.github.io/docs/concepts/interaction/actions) to allow for multi-select. Such context menu actions can be created using `createContextAction` as describe in the [context-menu section](https://launchmenu.github.io/docs/concepts/interaction/common-actions#context-menu-action). See the [execute section](https://launchmenu.github.io/docs/concepts/interaction/common-actions#custom-execute-handler) to learn how to specify behavior for when the user presses enter.

**Comment by TarVK** [comment]
I'm happy to provide more information if you have any specific questions later!

As you can see, things are not super straight forward, but due to the flexible structure of LM almost anything is possible! 
We would like to add more helper methods for common cases like these, but so far haven't been able to make much time to work on the project.

**Comment by faressoft** [comment]
Seems very close to what I am trying to build. Thank you. The only thing missing is the ability to disable filtering the returned list from the API since the list is already filtered based on complex criteria, not just the title. The user should be able to select one of the returned items.

Here is my current code:
```
// const resultCache = new SearchCache(createWordMenuItem);

export function createWordMenuItem(word: string): IMenuItem {
  return createStandardMenuItem({
    name: word,
    content: <Box>{word}</Box>,
    onExecute: ({ context }) => {},
  });
}

const search = async (query: IQuery, hook: IDataHook) => {
  // const search = query.search;

  const words = await (async () => {
    return ["item1 " + Math.random(), "item2 " + Math.random(), "item3 " + Math.random()];
  })();

  // const items = resultCache.getAll(words);
  const items = words.map(createWordMenuItem);

  return {
    children: searchAction.get(items),
  };
};

export default declare({
  info,
  settings,
  open({ context, onClose }) {
    const searchField = new TextField();
    const searchMenu = new SearchMenu(context, { search });
    const searchObserver = new Observer((h) => searchField.get(h)).listen((search) => searchMenu.setSearch(search));

    context.open(
      new UILayer(
        () => ({
          field: searchField,
          menu: searchMenu,
          searchable: false,
          onClose: () => {
            searchObserver.destroy();
            onClose();
          },
        }),
        { path: "Dictionary" }
      )
    );
  },
});
```

**Reply by TarVK** [reply, marked as answer]
Right, if you want to only search once your own applet has been opened (so not from the base search on LM's home screen) things become quite a bit easier actually! 

Your solution seems really close to correct (if I understand things correctly at least). You probably just don't want to use the SearchMenu which uses built-in prioritization. Instead you can use another menu type:

```ts
import {Field} from "model-react";
...

const search = async (query: string) => {
  const words = await (async () => {
    return ["item1 " + Math.random(), "item2 " + Math.random(), "item3 " + Math.random()];
  })();

  return words.map(createWordMenuItem);
};

...

    const searchField = new TextField();
    const searchMenu = new SearchMenu(context, { search });
    const items = new Field<IMenuItem[]>([]);
    const searchObserver = new Observer((h) => searchField.get(h)).listen((query) => {
       search(query).then(newItems=>items.set(newItems));
    });
    const menu = new ProxiedMenu(context, h => items.get(h));
    
    context.open( ... ); // Basically what you already had
```

So by using a [proxied menu](https://launchmenu.github.io/docs/concepts/ui/menu#proxiedmenu), no "smart" sorting is done. Alternatively you can also use the normal menu, which is more direct, but then you have to call `addItems` and `removeItems` which is a little more cumbersome than the `items.set()` we can use now. 

Additionally, I do recommend using the searchCache eventually. With the current setup, whenever a new item is created (for the same word) it also gets a new identifier. That means that if you select a word in the menu, then continue typing and the word is still there, it will still be deselected, because it's technically a new item (just representing the same thing). The SearchCache will make sure it's the same item and thus no unexpected deselection happens. At least that's my recollection about the search cache, it's been a while 😅

**Reply by faressoft** [reply]
@TarVK 

Perfect. Thank you.
